Geometry/MTDCommonData/plugins/DDMTDLinear.cc

```diff
@@ -0,0 +1,116 @@
+  double m_theta_obj;          //Theta angle to rotate volumes
+  std::vector<double> m_base;  //Base values - a 3d-point where the offset is calculated from
+                               //base is optional, if omitted base=(0,0,0)
+  std::pair<std::string, std::string> m_childNmNs;  //Child name
+                                                    //Namespace of the child
+};
+
+void DDMTDLinear::initialize(const DDNumericArguments& nArgs,
+                             const DDVectorArguments& vArgs,
+                             const DDMapArguments&,
+                             const DDStringArguments& sArgs,
+                             const DDStringVectorArguments&) {
+  m_n = int(nArgs["N"]);
+  m_startCopyNo = int(nArgs["StartCopyNo"]);
+  m_incrCopyNo = int(nArgs["IncrCopyNo"]);
+  m_theta = nArgs["Theta"];
+  m_phi = nArgs["Phi"];
+  m_delta = nArgs["Delta"];
+  m_base = vArgs["Base"];
+  m_phi_obj = nArgs["Phi_obj"];
+  m_theta_obj = nArgs["Theta_obj"];
+
+  LogDebug("DDAlgorithm") << "DDMTDLinear: Parameters for position"
+                          << "ing:: n " << m_n << " Direction Theta, Phi, Offset, Delta " << convertRadToDeg(m_theta)
+                          << " " << convertRadToDeg(m_phi) << " "
+                          << " " << convertRadToDeg(m_delta) << " Base " << m_base[0] << ", " << m_base[1] << ", "
+                          << m_base[2] << "Objects placement Phi_obj, Theta_obj " << convertRadToDeg(m_phi_obj) << " "
+                          << convertRadToDeg(m_theta_obj);
+
+  m_childNmNs = DDSplit(sArgs["ChildName"]);
+  if (m_childNmNs.second.empty())
+    m_childNmNs.second = DDCurrentNamespace::ns();
+
+  DDName parentName = parent().name();
+  LogDebug("DDAlgorithm") << "DDMTDLinear: Parent " << parentName << "\tChild " << m_childNmNs.first << " NameSpace "
+                          << m_childNmNs.second;
+}
+
+void DDMTDLinear::execute(DDCompactView& cpv) {
+  DDName mother = parent().name();
+  DDName ddname(m_childNmNs.first, m_childNmNs.second);
+  int copy = m_startCopyNo;
+
+  DDTranslation direction(sin(m_theta) * cos(m_phi), sin(m_theta) * sin(m_phi), cos(m_theta));
+
+  DDTranslation basetr(m_base[0], m_base[1], m_base[2]);
+
+  //rotation is in xy plane
+  double thetaZ = m_theta_obj - 0.5_pi;
+  double phiZ = m_phi_obj;
+  double thetaX = m_theta_obj;
+  double thetaY = m_theta_obj;
+  double phiX = m_phi_obj;
+  double phiY = m_phi_obj + 0.5_pi;
+
+  DDRotation rotation = DDRotation("Rotation");
+
+  if (!rotation) {
+    LogDebug("DDAlgorithm") << "DDMTDLinear: Creating a new "
+                            << "rotation for " << ddname;
+
+    rotation = DDrot("Rotation", DDcreateRotationMatrix(thetaX, phiX, thetaY, phiY, thetaZ, phiZ));
+  }
+
+  for (int i = 0; i < m_n; ++i) {
+    DDTranslation tran = basetr + (double(i) * m_delta) * direction;
+    cpv.position(ddname, mother, copy, tran, rotation);
```

Comment:
@ianna, @fabiocos Can this code be made more generic? Right now it is specialized for the old DD, which will make it harder to migrate.

Reply:
@cvuosalo there is already a DD4hep version in development by @icosivi, see https://github.com/fabiocos/cmssw/commit/a46c3d07e22420935abacb8b7c85e60b3d36d4d9
We would like to address this in a separate PR anyway

Reply:
BTW, this is moving along the port of @ianna https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/plugins/dd4hep/DDLinear.cc

Reply:
Sounds good.